### ISSUE TITLE
Fatal error: Uncaught TypeError: fclose(): supplied resource is not a valid stream resource

### ISSUE THREAD

**Issue by onet4** [elazar/flystream#2]
Hello,

I tried the example in readme and got the following error.

### Error
```bash
Fatal error: Uncaught TypeError: fclose(): supplied resource is not a valid stream resource in .../vendor/elazar/flystream/src/MemoryBuffer.php on line 40

TypeError: fclose(): supplied resource is not a valid stream resource in .../vendor/elazar/flystream/src/MemoryBuffer.php on line 40

Call Stack:
    0.0013     401440   1. {main}() .../scratch/flysystem-01.php:0
    0.2299    1156832   2. file_put_contents($filename = 'mem://foo/bar', $data = 'bay') .../scratch/flysystem-01.php:32
    0.2875    1161632   3. Elazar/Flystream/StreamWrapper->stream_close() .../scratch/flysystem-01.php:32
    0.2875    1161632   4. Elazar/Flystream/MemoryBuffer->close() .../vendor/elazar/flystream/src/StreamWrapper.php:162
    0.2875    1161632   5. fclose($stream = resource(50) of type (Unknown)) .../vendor/elazar/flystream/src/MemoryBuffer.php:40
```

### Code

```php
<?php
include( dirname( __DIR__ ) . '/vendor/autoload.php' );


use League\Flysystem\Filesystem;
use League\Flysystem\InMemory\InMemoryFilesystemAdapter;

$adapter = new InMemoryFilesystemAdapter();
$filesystem = new Filesystem($adapter);

/**
 * 2. Register the filesystem with Flystream and associate it with a
 *    custom protocol (e.g. 'mem').
 */

use Elazar\Flystream\FilesystemRegistry;
use Elazar\Flystream\ServiceLocator;

$registry = ServiceLocator::get(FilesystemRegistry::class);
$registry->register('mem', $filesystem);

/**
 * 3. Interact with the filesystem instance using the custom protocol.
 */

mkdir('mem://foo');

$file = fopen('mem://foo/bar', 'w');
fwrite($file, 'baz');
fclose($file);

file_put_contents('mem://foo/bar', 'bay');

$contents = file_get_contents('mem://foo/bar');
// or
$contents = stream_get_contents('mem://foo/bar');

if (file_exists('mem://foo/bar')) {
    rename('mem://foo/bar', 'mem://foo/baz');
    touch('mem://foo/bar');
}

$file = fopen('mem://foo/baz', 'r');
fseek($file, 2);
$position = ftell($file);
ftruncate($file, 0);
fclose($file);

$dir = opendir('mem://foo');
while (($entry = readdir($dir)) !== false) {
    echo $entry, PHP_EOL;
}
closedir($dir);

unlink('mem://foo/bar');
unlink('mem://foo/baz');

rmdir('mem://foo');

// These won't have any effect because Flysystem doesn't support them.
chmod('mem://foo', 0755);
chown('mem://foo', 'root');
chgrp('mem://foo', 'root');

/**
 * 4. Optionally, unregister the filesystem with Flystream.
 */
$registry->unregister('mem');
```

### Environment
- PHP: 8.0.9
- "elazar/flystream": "^0.3.0",
- "league/flysystem": "^3.0.8",
- "league/flysystem-memory": "^3.0"

**Comment by elazar** [comment]
@onet4 Thank you for this report! 😃 I am able to reproduce your issue. I'm in the process of trying to isolate the cause. I'll report back here when I have more to share. Please feel free to comment here again if you discover anything else that may be relevant to addressing this. Thanks again!

**Comment by elazar** [comment]
@onet4 I believe I have a fix for this issue. I would appreciate it if you could test it locally to ensure that it addresses your issue. The easiest way to apply the fix to your local environment is to run the following command.

```sh
composer require elazar/flystream:dev-memory-buffer-close-bug
```

Once you've finished, you can revert this change by running the command below.

```sh
composer require elazar/flystream:^0.3.0
```

Assuming that this fix works, I will need to release a new package version that includes it. When I do, I will reply to this issue with the new version number. If you'd like, I can tag you when I do so.

Lastly, I did find an issue with the example code in `README.md` once I applied this fix. I've included a diff of this change below so you can apply it to your local copy of the code you're trying to run. If it works for you, I'll add a commit to update `README.md` with this change.

```diff
 2 files changed, 12 insertions(+)

diff --git a/README.md b/README.md
index c48e7c6..8268e10 100644
--- a/README.md
+++ b/README.md
@@ -81,7 +81,7 @@ file_put_contents('mem://foo/bar', 'bay');

 $contents = file_get_contents('mem://foo/bar');
 // or
-$contents = stream_get_contents('mem://foo/bar');
+$contents = stream_get_contents(fopen('mem://foo/bar', 'r'));

 if (file_exists('mem://foo/bar')) {
     rename('mem://foo/bar', 'mem://foo/baz');
```

Once I made both of the above changes, I get this output from running the example code.

```
mem:/foo/baz
mem:/foo/bar
```

This segment of the example code is responsible for generating this output.

```php
$dir = opendir('mem://foo');
while (($entry = readdir($dir)) !== false) {
    echo $entry, PHP_EOL;
}
closedir($dir);
```

Please let me know if the above address your issues. Thanks again for your report.

**Comment by onet4** [comment]
I tried the `elazar/flystream:dev-memory-buffer-close-bug` branch and the example code produced 

```
mem:/foo/baz
mem:/foo/bar
```

which seems to be the expected behavior. 

So thank you for the fix.

**Comment by elazar** [comment]
@onet4 The fix for this issue has been released in 0.4.0. Thanks again for your report and for verifying the fix!
